forgerock-auth/src/main/java/org/forgerock/android/auth/devicebind/ApplicationPinDeviceAuthenticator.kt

```diff
@@ -0,0 +1,152 @@
+import java.io.FileNotFoundException
+import java.io.InputStream
+import java.io.OutputStream
+import java.security.KeyStore
+import java.security.PrivateKey
+import java.security.UnrecoverableKeyException
+import java.security.interfaces.RSAPublicKey
+import java.util.*
+import java.util.concurrent.Executors
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.atomic.AtomicReference
+
+/**
+ * Device Authenticator which use Application PIN to secure device cryptography keys
+ */
+open class ApplicationPinDeviceAuthenticator : CryptoAware, DeviceAuthenticator,
+    KeyStoreRepository {
+
+    @VisibleForTesting
+    internal lateinit var appPinAuthenticator: AppPinAuthenticator
+    internal lateinit var keyStore: KeyStoreRepository
+
+    private val pinSuffix = "_PIN"
+
+    protected lateinit var prompt: Prompt
+
+    @VisibleForTesting
+    internal val pinRef = AtomicReference<CharArray>()
+
+    internal val worker = Executors.newSingleThreadScheduledExecutor()
+
+    override suspend fun generateKeys(context: Context): KeyPair {
+        val pin = requestForCredentials()
+        yield() //Suspension point for timeout, we don't want to create the key if timeout
+        pinRef.set(pin)
+        //This allow a user to have an biometric key + application pin
+        val alias = appPinAuthenticator.getKeyAlias() + pinSuffix
+        val keyPair = appPinAuthenticator.generateKeys(context, pin)
+        //Clean up the password from memory after 1 second
+        worker.schedule({
+            pinRef.set(null)
+        }, 1, TimeUnit.SECONDS)
+        return KeyPair(keyPair.public as RSAPublicKey, keyPair.private, alias)
+    }
+
+    override suspend fun authenticate(context: Context): DeviceBindingStatus<PrivateKey> {
+
+        if (!appPinAuthenticator.exists(context)) {
+            return UnRegister()
+        }
+
+        var pin = pinRef.getAndSet(null)
+        pin?.let {
+            return getPrivateKey(context, it)
+        }
+        return try {
+            pin = requestForCredentials()
+            yield() //Suspension point for timeout, check to see if takes too long to get the pin
+            (getPrivateKey(context, pin))
+        } catch (e: OperationCanceledException) {
+            Abort()
+        }
+    }
+
+    override fun prompt(prompt: Prompt) {
+        this.prompt = prompt
+    }
+
+    override fun isSupported(context: Context): Boolean {
+        return true
+    }
+
+    /**
+     * Request for user credential
+     * @param fragmentActivity The current [FragmentActivity]
+     * @throws throw [OperationCanceledException] will return [Abort] status
+     */
+    open suspend fun requestForCredentials(fragmentActivity: FragmentActivity = InitProvider.getCurrentActivityAsFragmentActivity()): CharArray {
+        return suspendCancellableCoroutine { continuation ->
+
+            val existing =
+                fragmentActivity.supportFragmentManager.findFragmentByTag(ApplicationPinFragment.TAG)
+
```

Comment:
i think we can inline this

val existing =
                fragmentActivity.supportFragmentManager.findFragmentByTag(ApplicationPinFragment.TAG) as? ApplicationPinFragment

Reply:
Updated